$MirOS: ports/net/djbdns/patches/patch-src_okclient_c,v 1.4 2008/10/27 21:40:38 tg Exp $
--- src/okclient.c.orig	Sun Feb 11 21:11:23 2001
+++ src/okclient.c	Sat Mar 26 14:44:04 2016
@@ -2,24 +2,45 @@
 #include <sys/stat.h>
 #include "str.h"
 #include "ip4.h"
+#include "ip6.h"
+#include "byte.h"
 #include "okclient.h"
+#include "env.h"
 
-static char fn[3 + IP4_FMT];
+static char fn[3 + IP6_FMT];
 
-int okclient(char ip[4])
+int okclient(char ip[16])
 {
+  static int init_done = 0;
   struct stat st;
   int i;
+  char sep;
 
+  if (!init_done) {
+    if (env_get("OKCLIENT"))
+      init_done = 1;
+    else
+      init_done = 2;
+  }
+  if (init_done == 1)
+    return 1;
+
   fn[0] = 'i';
   fn[1] = 'p';
   fn[2] = '/';
-  fn[3 + ip4_fmt(fn + 3,ip)] = 0;
+  if (byte_equal(ip,12,V4mappedprefix)) {
+    fn[3 + ip4_fmt(fn + 3,ip+12)] = 0;
+    sep='.';
+  } else {
+    fn[3 + ip6_fmt(fn + 3,ip)] = 0;
+    sep=':';
+  }
 
   for (;;) {
+    if (!fn[3]) return 0;
     if (stat(fn,&st) == 0) return 1;
     /* treat temporary error as rejection */
-    i = str_rchr(fn,'.');
+    i = str_rchr(fn,sep);
     if (!fn[i]) return 0;
     fn[i] = 0;
   }
